### ISSUE TITLE
The output of `fuse_modules` and `fuse_modules_qat` is inconsistent

### ISSUE THREAD

**Issue by shaoyuyoung** [pytorch/pytorch#121470]
### 🐛 Describe the bug

When I try to fuse two modules (e.g., `Conv2d` and `BatchNorm2d`). I tried 1,000 times random input, finding that the output produced by `fuse_modules` is inconsistent with the output of the `original model` (1000/1000). I'm not sure whether it is caused by **Floating Point Value**.  So I try to use `fuse_modules_qat`, in this case, however, there is **no more inconsistency** (0/1000).

 As far as I know, fuse_modules folds weights of `BatchNorm2d`  into `Conv2d` instead of implementing **operator fusion**. Does this folding add up to errors?

Is `fuse_modules_qat` better than `fuse_modules` in this case? So what does `fuse_modules` do?

Thx in advance to solve my confusion.

Here is the code and output.

```python
import torch
from torch.ao.quantization.fuse_modules import fuse_modules, fuse_modules_qat
import torch.nn as nn
from tqdm import tqdm


class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.conv1 = nn.Conv2d(3, 64, 3, 1, 1)
        self.bn1 = nn.BatchNorm2d(64)

    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        return x


m = M()

modules_to_fuse = [['conv1', 'bn1']]

print(f'before operator fusion: \n{m}\n')
fused_m = fuse_modules(m.eval(), modules_to_fuse)
print(f'after operator fusion: \n{fused_m}\n')
fused_qat_m = fuse_modules_qat(m, modules_to_fuse)
print(f'after QAT operator fusion: \n{fused_qat_m}\n')

inconsistent_cnt = 0
qat_inconsistent_cnt = 0
for i in tqdm(range(1000)):
    random_input = torch.randn(1, 3, 224, 224)

    output_before_fusion = m(random_input)

    output_after_fusion = fused_m(random_input)

    output_after_qat_fusion = fused_qat_m(random_input)

    normal_are_equal = torch.allclose(output_before_fusion, output_after_fusion, rtol=1e-5, atol=1e-7)

    qat_are_equal = torch.allclose(output_before_fusion, output_after_qat_fusion, rtol=1e-5, atol=1e-7)

    if not normal_are_equal:
        inconsistent_cnt += 1

    if not qat_are_equal:
        qat_inconsistent_cnt += 1

print(f"The number of times the fusion results are inconsistent: {inconsistent_cnt}")
print(f"The number of times the qat fusion results are inconsistent: {qat_inconsistent_cnt}")
```

The output is  below:
```
before operator fusion: 
M(
  (conv1): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
)

after operator fusion: 
M(
  (conv1): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (bn1): Identity()
)

after QAT operator fusion: 
M(
  (conv1): ConvBn2d(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  )
  (bn1): Identity()
)

100%|███████████████████████████████████████| 1000/1000 [00:14<00:00, 71.14it/s]
The number of times the fusion results are inconsistent: 1000
The number of times the qat fusion results are inconsistent: 0
```

### Versions

Collecting environment information...
PyTorch version: 2.1.1
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.8.2
Libc version: glibc-2.27

Python version: 3.8.18 (default, Sep 11 2023, 13:40:15)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.15.0-213-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB

Nvidia driver version: 470.182.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] onnx==1.15.0
[pip3] onnxoptimizer==0.3.13
[pip3] onnxruntime==1.16.3
[pip3] torch==2.1.1
[pip3] torchaudio==2.1.1
[pip3] torchvision==0.16.1
[pip3] triton==2.1.0
[conda] Could not collect

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

**Comment by ZailiWang** [comment]
I tested with `atol=1e-6` tolerance setting and output of `fuse_modules` can pass the identity test.
I think such difference is generally acceptable.

**Comment by shaoyuyoung** [comment]
Thx for reply.

I don't think it's a serious error either. Just wonder why in this setting: `rtol=1e-5, atol=1e-7` can cause inconsistency of `fuse_modules` while `fuse_modules_qat` doesn't.

**Comment by jgong5** [comment]
> I don't think it's a serious error either. Just wonder why in this setting: `rtol=1e-5, atol=1e-7` can cause inconsistency of `fuse_modules` while `fuse_modules_qat` doesn't.

I guess that is because the computation out of `fuse_modules_qat` is identical to the original model?

**Comment by shaoyuyoung** [comment]
> > I don't think it's a serious error either. Just wonder why in this setting: `rtol=1e-5, atol=1e-7` can cause inconsistency of `fuse_modules` while `fuse_modules_qat` doesn't.
> 
> I guess that is because the computation out of `fuse_modules_qat` is identical to the original model?

Thank u for replying.

I will check it carefully

**Comment by shaoyuyoung** [comment]
So sorry for not focusing on this issue for a long time.

I am still confused by this loss of accuracy :(


Hi, @jerryzh168. What do you think about this?

**Comment by jerryzh168** [comment]
sorry for the late reply, I didn't get notifications. I think this is expected since fuse_modules is PTQ fusion where we fold the bn weights into conv, and fuse_modules_qat is fusing the conv bn to a ConvBn module for training that will reorder some of the operations so the conv bn module is more friendly to QAT

**Comment by shaoyuyoung** [comment]
Thank u :)
I can close this issue
